From afb1dbdb7a0ceda2678240eb0eda09400d8821ea Mon Sep 17 00:00:00 2001
From: zwx1252978 <zhangliang296@h-partners.com>
Date: Sat, 24 Jan 2026 17:03:13 +0800
Subject: [PATCH 2/3] ub: udma: bugfix related to log print

driver inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/IDAYM3
CVE: NA

-----------------------------------------------------------

This patch fix a bug about log print

Signed-off-by: Liang Zhang <zhangliang296@h-partners.com>
---
 src/urma/hw/udma/udma_u_ctl.c   | 24 ++++++++++--------------
 src/urma/hw/udma/udma_u_jetty.c | 12 ++++++------
 src/urma/hw/udma/udma_u_jfc.c   |  5 +++--
 src/urma/hw/udma/udma_u_jfs.c   |  4 ++--
 src/urma/hw/udma/udma_u_log.c   |  4 ++--
 src/urma/hw/udma/udma_u_log.h   |  3 ++-
 6 files changed, 25 insertions(+), 27 deletions(-)

diff --git a/src/urma/hw/udma/udma_u_ctl.c b/src/urma/hw/udma/udma_u_ctl.c
index cd4030a..d735684 100644
--- a/src/urma/hw/udma/udma_u_ctl.c
+++ b/src/urma/hw/udma/udma_u_ctl.c
@@ -8,10 +8,7 @@
  */
 
 #include <errno.h>
-#include <stdint.h>
 #include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
 #include <sys/types.h>
 #include "urma_private.h"
 #include "udma_u_common.h"
@@ -21,6 +18,7 @@
 #include "udma_u_jfr.h"
 #include "udma_u_jfc.h"
 #include "udma_u_jetty.h"
+#include "udma_u_ctrlq_tp.h"
 #include "udma_u_ctl.h"
 
 static void udma_u_uninit_queue_buf(struct udma_u_jetty_queue *q)
@@ -84,7 +82,7 @@ static int udma_u_init_jfs_queue_buf(struct udma_u_jetty_queue *q,
 
 	q->wrid = (uintptr_t *)malloc(q->baseblk_cnt * sizeof(uint64_t));
 	if (q->wrid == NULL) {
-		UDMA_LOG_ERR("failed to alloc jfs STARS buffer for wrid!\n");
+		UDMA_LOG_ERR("failed to alloc stars buffer for wrid!\n");
 		return ENOMEM;
 	}
 
@@ -341,13 +339,8 @@ static int udma_u_verify_jfs_param_ex(urma_context_t *ctx, struct udma_u_jfs_cfg
 	urma_device_attr_t *attr;
 	urma_jfs_cfg_t *jfs_cfg;
 
-	if (ctx == NULL) {
-		UDMA_LOG_ERR("urma ctx is null.\n");
-		return EINVAL;
-	}
-
-	if (cfg_ex == NULL) {
-		UDMA_LOG_ERR("cfg ex is null.\n");
+	if (ctx == NULL || cfg_ex == NULL) {
+		UDMA_LOG_ERR("urma ctx or cfg ex is null.\n");
 		return EINVAL;
 	}
 
@@ -438,7 +431,7 @@ static int udma_u_create_jetty_check_cap(urma_context_t *ctx, urma_jetty_cfg_t *
 
 	if ((jfs_cfg->max_inline_data != 0 && jfs_cfg->max_inline_data > cap->max_jfs_inline_len) ||
 	    (jfr_cfg->depth == 0 || jfr_cfg->depth > cap->max_jfr_depth) ||
-	    (jfs_cfg->max_sge > cap->max_jfs_sge || jfs_cfg->max_rsge > cap->max_jfs_rsge || \
+	    (jfs_cfg->max_sge > cap->max_jfs_sge || jfs_cfg->max_rsge > cap->max_jfs_rsge ||
 	     jfr_cfg->max_sge > cap->max_jfr_sge)) {
 		UDMA_LOG_ERR("jetty cfg out of range, jfs_depth:%u, max_jfs_depth: %u, " \
 			"inline_data:%u, max_jfs_inline_len: %u, jfr_depth:%u, max_jfr_depth: %u, " \
@@ -696,7 +689,7 @@ static int udma_u_jfr_ops_ex(urma_context_t *ctx, urma_user_ctl_in_t *in,
 		if (jfr == NULL)
 			return EFAULT;
 
-		memcpy((void *)out->addr, &jfr, sizeof(urma_jfr_t *));
+		(void)memcpy((void *)out->addr, &jfr, sizeof(urma_jfr_t *));
 		atomic_fetch_add(&ctx->ref.atomic_cnt, 1);
 	} else {
 		if (!udma_u_user_ctl_check_param(in->addr, in->len, (uint32_t)sizeof(urma_jfr_t *), op))
@@ -1086,6 +1079,8 @@ static int udma_u_ctrlq_query_tp_sport(urma_context_t *ctx, urma_user_ctl_in_t *
 	if (ret)
 		UDMA_LOG_ERR("failed to query tp sport, ret: %d.\n", ret);
 
+	in->opcode = UDMA_U_USER_CTL_QUERY_TP_SPORT;
+
 	return ret;
 }
 
@@ -1170,7 +1165,8 @@ int udma_u_user_ctl(urma_context_t *ctx, urma_user_ctl_in_t *in,
 		return URMA_EINVAL;
 	}
 
-	if (in->opcode >= UDMA_U_USER_CTL_MAX) {
+	if (in->opcode >= ARRAY_SIZE(g_udma_u_user_ctl_ops) ||
+	    g_udma_u_user_ctl_ops[in->opcode] == NULL) {
 		UDMA_LOG_ERR("invalid opcode: 0x%x.\n", in->opcode);
 		return URMA_ENOPERM;
 	}
diff --git a/src/urma/hw/udma/udma_u_jetty.c b/src/urma/hw/udma/udma_u_jetty.c
index b13983d..67e8a5f 100644
--- a/src/urma/hw/udma/udma_u_jetty.c
+++ b/src/urma/hw/udma/udma_u_jetty.c
@@ -370,7 +370,7 @@ urma_status_t udma_u_unimport_jetty(urma_target_jetty_t *target_jetty)
 
 	if (target_jetty->trans_mode == URMA_TM_RC &&
 	    target_jetty->tp.tpn != INVALID_TPN) {
-		UDMA_LOG_ERR("the RC target jetty is still being used, id = %u.\n",
+		UDMA_LOG_ERR("the rc target jetty is still being used, id = %u.\n",
 			     target_jetty->id.id);
 		return URMA_FAIL;
 	}
@@ -426,14 +426,14 @@ urma_status_t udma_u_unbind_jetty(urma_jetty_t *jetty)
 	int ret;
 
 	if (tjetty == NULL || udma_jetty->sq.tjetty == NULL) {
-		UDMA_LOG_ERR("The Jetty not bind a remote Jetty, id = %u.\n",
+		UDMA_LOG_ERR("the jetty not bind a remote jetty, id = %u.\n",
 			     jetty->jetty_id.id);
 		return URMA_EINVAL;
 	}
 
 	if (udma_jetty->sq.trans_mode != URMA_TM_RC ||
 	    tjetty->trans_mode != URMA_TM_RC) {
-		UDMA_LOG_ERR("The transmode of Jetty or tJetty is not rc.\n");
+		UDMA_LOG_ERR("the transmode of jetty or tjetty is not rc.\n");
 		return URMA_EINVAL;
 	}
 
@@ -586,7 +586,7 @@ urma_target_jetty_t *udma_u_import_jetty_ex(urma_context_t *ctx,
 	struct udma_u_target_jetty *tjetty;
 
 	if (rjetty->type != URMA_JETTY && rjetty->type != URMA_JETTY_GROUP) {
-		UDMA_LOG_ERR("The Jetty type %u cannot imported jetty ex.\n",
+		UDMA_LOG_ERR("the jetty type %u cannot imported jetty ex.\n",
 			     rjetty->type);
 		return NULL;
 	}
@@ -631,7 +631,7 @@ urma_status_t udma_u_bind_jetty_ex(urma_jetty_t *jetty,
 
 	if (udma_jetty->sq.trans_mode != URMA_TM_RC ||
 	    tjetty->trans_mode != URMA_TM_RC) {
-		UDMA_LOG_ERR("The transmode of Jetty or tJetty in exp is not rc.\n");
+		UDMA_LOG_ERR("the transmode of jetty or tjetty in exp is not rc.\n");
 		return URMA_EINVAL;
 	}
 
@@ -642,7 +642,7 @@ urma_status_t udma_u_bind_jetty_ex(urma_jetty_t *jetty,
 	}
 
 	if (udma_jetty->sq.tjetty != NULL) {
-		UDMA_LOG_ERR("The rc Jetty has bound a remote Jetty in exp,"
+		UDMA_LOG_ERR("the rc jetty has bound a remote jetty in exp,"
 			     "jetty_id = %u.\n", jetty->jetty_id.id);
 		return URMA_EEXIST;
 	}
diff --git a/src/urma/hw/udma/udma_u_jfc.c b/src/urma/hw/udma/udma_u_jfc.c
index 4b51c97..7455542 100644
--- a/src/urma/hw/udma/udma_u_jfc.c
+++ b/src/urma/hw/udma/udma_u_jfc.c
@@ -133,6 +133,7 @@ err_alloc_sw_db:
 	udma_u_delete_cq(&jfc->cq);
 err_create_cq:
 	free(jfc);
+
 	return NULL;
 }
 
@@ -194,7 +195,7 @@ struct udma_cr_status {
 		return JFC_OK;
 	}
 
-	UDMA_LOG_ERR("cqe_status (%u) substatus (%u) is invalid.",
+	UDMA_LOG_ERR("cqe_status (%u) substatus (%u) is invalid.\n",
 		     src_status, substatus);
 
 	return JFC_POLL_ERR;
@@ -666,7 +667,7 @@ void udma_u_clean_jfc(struct urma_jfc *jfc, uint32_t jetty_id)
 			break;
 	}
 
-	while ((int) --pi - (int) cq->ci >= 0) {
+	while ((int)--pi - (int)cq->ci >= 0) {
 		cqe = (struct udma_u_jfc_cqe *)get_u_buf_entry(cq, pi);
 		UDMA_FROM_DEVICE_BARRIER();
 		local_id = (cqe->local_num_h << UDMA_SRC_IDX_SHIFT) | cqe->local_num_l;
diff --git a/src/urma/hw/udma/udma_u_jfs.c b/src/urma/hw/udma/udma_u_jfs.c
index 5e7fcd7..c577bf8 100644
--- a/src/urma/hw/udma/udma_u_jfs.c
+++ b/src/urma/hw/udma/udma_u_jfs.c
@@ -313,7 +313,7 @@ static bool udma_check_sge_num_and_opcode(urma_opcode_t opcode, struct udma_u_je
 		*udma_opcode = UDMA_OPCODE_NOP;
 		goto default_sge_num;
 	default:
-		UDMA_LOG_ERR("Invalid opcode :%u\n", (uint8_t)opcode);
+		UDMA_LOG_ERR("invalid opcode :%u\n", (uint8_t)opcode);
 		return true;
 	}
 
@@ -762,7 +762,7 @@ static urma_status_t udma_set_sqe(struct udma_jfs_sqe_ctl *wqe_ctl,
 
 	wqe_ctl->rmt_jetty_type = (uint8_t)(tjetty->type);
 	if (udma_parse_jfs_wr(wqe_ctl, wr, sq, wqe_info, tjetty) != 0) {
-		UDMA_LOG_ERR("Failed to parse wr\n");
+		UDMA_LOG_ERR("failed to parse wr\n");
 		return URMA_EINVAL;
 	}
 
diff --git a/src/urma/hw/udma/udma_u_log.c b/src/urma/hw/udma/udma_u_log.c
index 6a660a5..d1143f0 100644
--- a/src/urma/hw/udma/udma_u_log.c
+++ b/src/urma/hw/udma/udma_u_log.c
@@ -62,7 +62,7 @@ void udma_getenv_log_level(void)
 
 	if (strnlen(level_str, UDMA_LOG_LEVEL_ENV_MAX_BUF_LEN) >=
 	    UDMA_LOG_LEVEL_ENV_MAX_BUF_LEN) {
-		UDMA_LOG_ERR("Invalid parameter: log level str.");
+		UDMA_LOG_ERR("Invalid parameter: log level str.\n");
 		return;
 	}
 
@@ -87,7 +87,7 @@ int udma_vlog(const char *function, int line, enum udma_vlog_level level, const
 		return ret;
 
 	ret = vsnprintf(logmsg, MAX_LOG_LEN, newformat, va);
-	if (ret == -1) {
+	if (ret == -1 || ret >= (int)sizeof(newformat)) {
 		(void)printf("logmsg size exceeds MAX_LOG_LEN size :%d.\n", MAX_LOG_LEN);
 		return ret;
 	}
diff --git a/src/urma/hw/udma/udma_u_log.h b/src/urma/hw/udma/udma_u_log.h
index 4ce3c95..0689069 100644
--- a/src/urma/hw/udma/udma_u_log.h
+++ b/src/urma/hw/udma/udma_u_log.h
@@ -9,6 +9,7 @@
 
 #ifndef __UDMA_U_LOG_H__
 #define __UDMA_U_LOG_H__
+
 #include <pthread.h>
 #include <stdbool.h>
 
@@ -33,7 +34,7 @@ bool udma_log_drop(enum udma_vlog_level level);
 void udma_getenv_log_level(void);
 void udma_log(const char *function, int line, enum udma_vlog_level level,
 	      const char *format, ...) __attribute__((format(printf, 4, 5)));
-enum udma_vlog_level udma_log_get_level_from_string(const char *level_string);
+enum udma_vlog_level udma_log_get_level_from_string(const char* level_string);
 
 #define UDMA_LOG(l, ...)								\
 	do {										\
-- 
2.28.0.windows.1

